src/XMakeBuildEngine/Evaluation/MSBuildExtensionsPathHelper.cs

```diff
@@ -0,0 +1,164 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.Evaluation
+{
+
+#if FEATURE_EXTENSIONS_PATH_SEARCH_PATHS
+    internal static class MSBuildExtensionsPathHelper
+    {
+        /// <summary>
+        /// Directory on OSX, used for lookup of project files imported from the property MSBuildExtensionsPath*
+        /// </summary>
+        private static readonly string s_osxExternalXBuildDir = "/Library/Frameworks/Mono.framework/External/xbuild";
+
+        /// <summary>
+        /// Container class for all the extensions paths lists
+        /// Lazy<> to make access to the extensions path lists, lazy and thread safe.
+        /// </summary>
+        private static Lazy<ExtensionsPathsLists> s_pathsLists = new Lazy<ExtensionsPathsLists>(true);
+
+        public static List<string> GetFallbackExtensionsPathsFor(MSBuildExtensionsPathReferenceKind refKind)
+        {
+            switch(refKind) {
+                case MSBuildExtensionsPathReferenceKind.Default:
+                    return s_pathsLists.Value.Default;
+                case MSBuildExtensionsPathReferenceKind.Path32:
+                    return s_pathsLists.Value.Path32;
+                case MSBuildExtensionsPathReferenceKind.Path64:
+                    return s_pathsLists.Value.Path64;
+                case MSBuildExtensionsPathReferenceKind.None:
+                    return new List<string> ();
+                default:
+                    // new enum value added?
+                    throw new NotImplementedException();
+            }
+        }
+
+        private static List<string> ComputeFallbackExtensionsPathsFor(MSBuildExtensionsPathReferenceKind refKind)
+        {
+            var extnPaths = new List<string>();
+            var refKindAsString = ReferenceKindToString(refKind);
+            var envVar = Environment.GetEnvironmentVariable(refKindAsString + "SearchPaths");
+
+            var allPaths = String.Join(Path.PathSeparator.ToString(),
+                                        //FIXME: IsOSX returns false :/
+                                        NativeMethodsShared.IsUnix ? s_osxExternalXBuildDir : String.Empty,
```

Comment:
Did you mean IsUnixLike? Either way with the solution I proposed hopefully this should go away and be completely generic here.